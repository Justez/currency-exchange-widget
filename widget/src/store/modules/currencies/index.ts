//@ts-ignore
import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencies';

export const actions = createActions(
  'SET_SELECTED_CURRENCY',
  'FLIP_SELECTED_CURRENCIES',
  { prefix: NAMESPACE },
);
