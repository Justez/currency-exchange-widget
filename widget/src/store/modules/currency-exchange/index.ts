//@ts-ignore
import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencyRates';

export const actions = createActions(
  'START_POLLING_CURRENCY_RATES',
  'STOP_POLLING_CURRENCY_RATES',
  'GET_CURRENCY_RATES',
  'GET_CURRENCY_RATES_REQUEST',
  'GET_CURRENCY_RATES_SUCCESS',
  'GET_CURRENCY_RATES_FAILURE',
  'SET_CURRENCY_RATES',
  { prefix: NAMESPACE },
);
