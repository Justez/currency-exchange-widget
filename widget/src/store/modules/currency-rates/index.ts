import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencyRates';

export const actions = createActions(
  'FLIP_RATES',
  'GET_CURRENCY_RATES',
  'GET_CURRENCY_RATES_REQUEST',
  'GET_CURRENCY_RATES_SUCCESS',
  'GET_CURRENCY_RATES_FAILURE',
  'SET_CURRENCY_RATES',
  'RESET_RATES',
  { prefix: NAMESPACE },
);
