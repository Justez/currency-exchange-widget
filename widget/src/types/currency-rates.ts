import { CurrencyMap } from 'types';

export type CurrencyRateResponse = {
  data: {
    base: string;
    rates: {
      [key in CurrencyMap]: number;
    };
  }
}

export type CurrencyRateError = {
    error: boolean;
    status: number;
    message: string;
    description: string;
}