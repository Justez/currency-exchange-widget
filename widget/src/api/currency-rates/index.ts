import axios from 'axios'

import { Currencies,CurrencyRateResponse, CurrencyRateError } from 'types';

const URL = 'https://openexchangerates.org/api/latest.json';

export const getCurrencyRates = (obj: Currencies) => axios.get(URL, {
    params: {
        app_id: '2b3e9170d55c46fead589f4fdeac1fb8',
        // base: obj.out.toUpperCase(),
    }
})
    .then((res: CurrencyRateResponse) => res.data)
    .catch((err: CurrencyRateError) => err);