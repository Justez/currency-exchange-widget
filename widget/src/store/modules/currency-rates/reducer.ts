import { handleActions } from 'redux-actions';
import { assocPath } from 'ramda';

import { CurrencyRates } from 'types';

import { actions } from '.';

export type DefaultState = CurrencyRates;

export const defaultState: DefaultState = {
  rate: '',
  reverse: '',
};

type Payload = any;

export const initializedState = {};

const reducer = handleActions<DefaultState, Payload>({
  [actions.setCurrencyRates.toString()]: (state, { payload }) =>
    assocPath([], { ...state, ...payload }, state),
  [actions.flipRates.toString()]: ((state) =>
    assocPath([], { rate: state.reverse, reverse: state.rate }, state)),
  [actions.resetRates.toString()]: ((state) =>
    assocPath([], defaultState, state)),
},
  defaultState,
);

export default reducer;
