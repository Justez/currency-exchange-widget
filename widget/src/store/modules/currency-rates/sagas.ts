import {
  all,
  takeLatest,
  put,
  select,
  race,
  call,
  take,
} from 'redux-saga/effects';
import delay from '@redux-saga/delay-p'

import { actions as currencyActions } from 'store/modules/currencies';
import { getSelectedCurrencies } from 'store/modules/currencies/selectors'
import { getChecksum } from './selectors'
import { actions } from 'store/modules/currency-rates';
import { getCurrencyRates } from 'api/currency-rates'

function* getCurrencyRatesFlow() {
  try {
    const checksum = yield select(getChecksum)
    yield put({ type: actions.getCurrencyRatesRequest.toString() })
    const selectedCurrencies = yield select(state => getSelectedCurrencies(state))

    const { response } = yield race({ response: call(getCurrencyRates, selectedCurrencies, checksum), cancel: take(actions.registerChecksum.toString()) })
    if (response) { // TODO parse data forward
      yield put({ type: actions.setCurrencyRates.toString(), payload: response.data })
      yield put({ type: actions.getCurrencyRatesSuccess.toString() })
    } else {
      yield put({ type: actions.getCurrencyRatesFailure.toString() })
      yield put({ type: actions.getCurrencyRates.toString() })
    }
    
    yield delay(50000)
    yield put({ type: actions.getCurrencyRates.toString() })
  } catch (error) {
  }
}

function* pollRates() {
  try {
    yield put({ type: actions.getCurrencyRates.toString() })
  } catch (error) {
  }
}

function* registerCheckSum() {
  try {
    yield put({ type: actions.registerChecksum.toString() })
  } catch (error) {
  }
}

function* watchCurrencyPairChanged() {
  yield takeLatest([currencyActions.setSelectedCurrency, currencyActions.flipSelectedCurrencies], registerCheckSum)
}

export default function* saga() {
  yield all([
    watchCurrencyPairChanged(),
    takeLatest(actions.registerChecksum, pollRates),
    takeLatest(actions.getCurrencyRates, getCurrencyRatesFlow)
  ]);
}
