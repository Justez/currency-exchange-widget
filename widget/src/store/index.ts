import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';

import createRootReducer from './root-reducer';
import sagas from './root-saga';

const configureStore = () => {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares = [
    sagaMiddleware,
  ];

  const instance = createStore(
    createRootReducer(),
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  sagaMiddleware.run(sagas);

  return instance;
};

export type State = ReturnType<typeof store.getState>;

const store = configureStore();

export default store;
