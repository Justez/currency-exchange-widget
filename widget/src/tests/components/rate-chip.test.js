import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import RateChip from 'components/rate-chip';
import { actions as currencyRatesActions } from 'store/modules/currency-rates';

import store from 'store';

const component = (
    <Provider store={store}>
        <RateChip />
    </Provider>
)

let wrapper, chip, label;

describe('RateChip component', () => {
    const load = () => {
        wrapper = mount(component);
        chip = wrapper.find('div[role="button"]');
    };

    beforeAll(() => load());

    it('is rendered', () => {
        expect(chip.exists()).toBeTruthy();
    });

    it('shows unavailable', () => {
        label = chip.find('span[test-id="rate-display-error"]')
        expect(label.props().children).toBe('unavailable');
    })

    it('renders unavailable state correctly', () => {
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('shows label with rates', () => {
        store.dispatch(currencyRatesActions.setCurrencyRates({ rate: 1.22, reverse: 0.88 }));
        load();
        label = chip.find('[test-id="rate-display"]').at(0);
        expect(label.props().children[0]).toBe('1');
        expect(label.props().children[3]).toBe(1.22);
    });
});
