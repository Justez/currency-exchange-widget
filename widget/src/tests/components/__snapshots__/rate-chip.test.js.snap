// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateChip component renders unavailable state correctly 1`] = `
<div
  className="MuiButtonBase-root MuiChip-root makeStyles-root-1 MuiChip-outlined MuiChip-clickable"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root"
    focusable="false"
    style={
      Object {
        "height": "24px",
        "margin": "4px",
        "width": "24px",
      }
    }
    viewBox="0 0 24 24"
  >
    <path
      d="M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
    />
  </svg>
  <span
    className="MuiChip-label"
  >
    <span
      className="MuiTypography-root makeStyles-label-2 MuiTypography-caption"
      test-id="rate-display-error"
    >
      unavailable
    </span>
  </span>
</div>
`;
